ivre/db/sql/__init__.py

```diff
@@ -809,13 +811,82 @@ def __migrate_schema_9_10(self):
         )
         return len(failed)
 
-    def __migrate_schema_10_11(self):
+    def _migrate_schema_10_11(self):
         """Converts a record from version 10 to version 11. Version 11 changes
 the way IP addresses are stored.
 
         """
         raise NotImplementedError
 
+    def _migrate_schema_11_12(self):
+        """Converts a record from version 11 to version 12. Version 12 changes
+the structured output for fcrdns and rpcinfo script.
+
+        """
+        failed = set()
+        req = (select([self.tables.scan.id,
```

Comment:
You could maybe merge the two `select` requests and avoid some code duplication.

```Python
req = (select([self.tables.scan.id,
    self.tables.script.name,
    self.tables.script.port,
    self.tables.script.output,
    self.tables.script.data])
.select_from(join(join(self.tables.scan, self.tables.port),
    self.tables.script))
.where(and_(self.tables.scan.schema_version == 11,
    or_(self.tables.script.name == "fcrdns",
        self.tables.script.name == "rpcinfo")))
for rec in self.db.execute(req):
    try:
        if 'fcrdns' in rec.data:
            # Since dots in keys are converted to
            # underscores, we need to convert them
            # back. It's OK to do so because underscores
            # cannot be used in FQDNs.
            data = [
                dict(result, name=name.replace('_', '.'))
                for name, result in viewitems(rec.data["fcrdns"])
            ]
        else:
            data = xmlnmap.change_rpcinfo(rec.data["rpcinfo"])
        except Exception:
            utils.LOGGER.warning("Cannot migrate host %r", rec.id,
                exc_info=True)
            failed.add(rec.id)
        else:
            if data:
                self.db.execute(
                    update(self.tables.script)
                    .where(and_(self.tables.script.port == rec.port,
                        self.tables.script.name == rec.name))
                    .values(data={rec.name: data})
                    )
```

(I did not test the code above)

Reply:
I have used `.in_()` rather than `or_()`

Reply:
I have used `.in_()` instead of `or_()` but that's the same idea.